core/blockchain.go

```diff
@@ -1212,108 +1448,268 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
+								return i, events, coalescedLogs, err
+							}
+							bc.AddFinalizedTrades(finalizedTx.TxHash, finalizedTrades)
+						}
+					}
+				}
+				//check
+				if tradingState != nil && tradingService != nil {
+					gotRoot := tradingState.IntermediateRoot()
+					expectRoot, _ := tradingService.GetTradingStateRoot(block, author)
+					parentRoot, _ := tradingService.GetTradingStateRoot(parent, parentAuthor)
+					if gotRoot != expectRoot {
+						err = fmt.Errorf("invalid XDCx trading state merke trie got : %s , expect : %s ,parent : %s", gotRoot.Hex(), expectRoot.Hex(), parentRoot.Hex())
+						bc.reportBlock(block, nil, err)
+						return i, events, coalescedLogs, err
+					}
+					log.Debug("XDCX Trading State Root", "number", block.NumberU64(), "parent", parentRoot.Hex(), "nextRoot", expectRoot.Hex())
+				}
+				if lendingState != nil && tradingState != nil {
+					gotRoot := lendingState.IntermediateRoot()
+					expectRoot, _ := lendingService.GetLendingStateRoot(block, author)
+					parentRoot, _ := lendingService.GetLendingStateRoot(parent, parentAuthor)
+					if gotRoot != expectRoot {
+						err = fmt.Errorf("invalid lending state merke trie got : %s , expect : %s , parent :%s ", gotRoot.Hex(), expectRoot.Hex(), parentRoot.Hex())
+						bc.reportBlock(block, nil, err)
+						return i, events, coalescedLogs, err
+					}
+					log.Debug("XDCX Lending State Root", "number", block.NumberU64(), "parent", parentRoot.Hex(), "nextRoot", expectRoot.Hex())
+				}
+			}
 		}
+		feeCapacity := state.GetTRC21FeeCapacityFromStateWithCache(parent.Root(), statedb)
 		// Process block using the parent state as reference point.
-		t0 := time.Now()
-		receipts, logs, usedGas, err := bc.processor.Process(block, state, bc.vmConfig)
-		t1 := time.Now()
+		receipts, logs, usedGas, err := bc.processor.Process(block, statedb, tradingState, bc.vmConfig, feeCapacity)
 		if err != nil {
 			bc.reportBlock(block, receipts, err)
-			return it.index, events, coalescedLogs, err
+			return i, events, coalescedLogs, err
 		}
 		// Validate the state using the default validator
-		if err := bc.Validator().ValidateState(block, parent, state, receipts, usedGas); err != nil {
+		err = bc.Validator().ValidateState(block, parent, statedb, receipts, usedGas)
+		if err != nil {
 			bc.reportBlock(block, receipts, err)
-			return it.index, events, coalescedLogs, err
+			return i, events, coalescedLogs, err
 		}
-		t2 := time.Now()
-		proctime := time.Since(start)
-
+		proctime := time.Since(bstart)
 		// Write the block to the chain and get the status.
-		status, err := bc.WriteBlockWithState(block, receipts, state)
-		t3 := time.Now()
+		status, err := bc.WriteBlockWithState(block, receipts, statedb, tradingState, lendingState)
 		if err != nil {
-			return it.index, events, coalescedLogs, err
+			return i, events, coalescedLogs, err
+		}
+		if bc.chainConfig.XDPoS != nil {
```

Comment:
What's the main purpose of this while block of code? It seems it just does the logging?